### PR TITLE
Fix issue 676: make bind-by-fn require all attrs to have values

### LINKED ISSUE #676: [Bug]: bind-by-fn computes wrong result if not all symbols have values
### What version of Datahike are you using?

0.6.1659

### What version of Java are you using?

openjdk version "20.0.1" 2023-04-18

### What operating system are you using?

Ubuntu 22

### What database EDN configuration are you using?

```
(d/db-with [{:db/id 1, :name  "Ivan",  :age   15}
                           {:db/id 2, :name  "Petr",  :age   22, :height 240, :parent 1}
                           {:db/id 3, :name  "Slava", :age   37, :parent 2}
                           {:db/id 4, :name  "Ivan",  :age   22}])
```

### Describe the bug

When using clauses of type `bind-by-fn`, such as `[(get-else $ ?e :height 300) ?height]`, all symbols that are function arguments must have some values in the context. In this example, we have the symbols `?e` and `$` (the database). If the symbol `?e` is not a constant or part of a relation in the context, the function `datahike.query/bind-by-fn` will silently compute the wrong result.

### What is the expected behaviour?

The function `bind-by-fn` must not produce the wrong result if symbols are missing.

### How can the behaviour be reproduced?

Currently, the query engine evaluates the clauses in the order they are listed in the expression. The following assertion that is part of `datahike.test.query-fns-test/test-query-fns` works because the clause `[?e :age ?age]` will make sure that the context will contain the symbol `?e` before the `get-else` clause is processed:

```clj
(is (= #{[1 15 300] [2 22 240] [3 37 300] [4 22 300]}
             (d/q '[:find ?e ?age ?height
                    :where [?e :age ?age]
                    [(get-else $ ?e :height 300) ?height]] db)))
```

By changing the order of the clauses, the query engine will compute the wrong result and the unit test will fail:

```clj
(is (= #{[1 15 300] [2 22 240] [3 37 300] [4 22 300]}
             (d/q '[:find ?e ?age ?height
                    :where
                    [(get-else $ ?e :height 300) ?height]
                    [?e :age ?age]] db)))
```

### PR DESCRIPTION
**Pull request by jonasseglare**
#### SUMMARY
Fixes #676 using the following changes:

* Rewrite `datahike.query/bind-by-fn` to return nil if not all symbols in the function call can be bound to values.
* Add function `datahike.tools/resolve-clauses`.
* Replace all occurrences in of `(reduce -resolve-clause ...)`  by `(dt/resolve-clauses ...)` to handle the `nil` returned by `datahike.query/bind-by-fn`.
* Add unit test that demonstrates that the bug has been fixed.

*Discussion about the solution to issue 676:*
The code in `bind-by-fn` is currently written in such a way that it assumes all symbols in the function call to be known in the context when the clause is evaluated but there are not checks to ensure that this is actually the case. Maybe there is a way to make it possible for symbols to be unknown but that would probably require a substantial rewrite of `bind-by-fn`. An easier way to address this problem is to make `bind-by-fn` return `nil`, instead of the new context, if not all symbols are known. However, this `nil` will consequently be returned by `-resolve-clause` and needs to be handled. We handle that `nil` by replacing all occurences of `(reduce -resolve-clause ...)` by `(dt/resolve-clauses ...)` to handle the nil and retry failed clauses in subsequent iterations.

##### Bugfix
- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [ ] Architecture Decision Record added if design changes necessary
- [ ] Formatting checked

### CODE REVIEW

File: src/datahike/query.cljc [hunk 1]
```diff
@@ -750,43 +750,55 @@
...
+          [context production] (rel-prod-by-attrs context attrs)
+          symbols-with-values (into #{}
+                                    (mapcat keys)
+                                    [(:attrs production)
+                                     (:consts context)
+                                     (:sources context)])]
+      ;; Currently, we can only evaluate this clause if all variables
+      ;; in the function call are bound. If not, we return nil which
+      ;; is handled by `datahike.tools/resolve-clauses`.
+      (when (every? symbols-with-values attrs)
```
Comment (jonasseglare) [inline src/datahike/query.cljc:773]: This `(when ...)` is the only relevant change in this function. Everything else is pretty much change of indentation and binding `attrs` and `symbols-with-values`.

Comment (whilo) [reply src/datahike/query.cljc:773]: That makes sense. I think the indentation changed though, which makes it harder to see changes and also will trigger our format checker.

**Review by whilo — CHANGES_REQUESTED** [review]
Should we throw an exception when not all arguments to a function are bound? @jsmassa and I discussed doing some static analysis (before running the query) to determine whether variables are bound or clause reordering was necessary. Maybe I misunderstand though.

**Review by whilo — APPROVED** [review]
Thanks, after our discussion on slack this is good to go!

**Pull request merged**